tellor-io/TellorMiner
# improve tracker coverage

# How to Run These Tests and view coverage
0. ensure go is installed and `go get -d .` has been run in TellorMiner root to install all dependencies, as discussed [here](https://tellor.readthedocs.io/en/latest/MinerSetupFromSource/#run-tellorminer-from-source)
1. In the TellorMiner project root run
    ```bash
    $ ./runPkgTest tracker
    ```
    
    All tests should pass,  a `trackercoverage.out` file should be created, and the console should log out the coverage percentages of all tested files.
2. To view the coverage report locally, run
    ```bash
    $ go tool cover -html=trackercoverage.out
   ```
    A tab should open in the default browser displaying a visual coverage report for all tested files.

3. Coverage reports and statistics are now available remotely on [Coveralls](https://coveralls.io/github/tellor-io/TellorMiner), thanks to the [github action CI workflow](https://github.com/tellor-io/TellorMiner/commit/7f9bfe48b1ccea6d9760281f8929d552b9d15484)

cc @themandalore and @JGcarv for acceptance testing
**note** locally I was using golang v1.13.3

## Future Work to Consider

Ideally, every package in this repo is covered with functioning tests to some degree. Eventually, all tests across packages should be run using the go module test infrastructure, as opposed to simply calling package tests. This should allow coverage logs to be combined to a single file and for  a seamlessly combined coverage report for the entire project, but package-level coverage can be submitted for now as tests are updated

---
remaining tasks to facilitate increasing full coverage:
- [x] stub out http requests to make testing morepredictable
- [x] create a mock rpc block header
